activerecord/lib/active_record/core.rb

```diff
@@ -209,7 +209,7 @@ def find_by(*args) # :nodoc:
 
         keys = hash.keys
 
-        statement = cached_find_by_statement(keys) { |params|
+        statement = cached_find_by_statement("#{table_name}#{keys}") { |params|
```

Comment:
The `table_name=` method already clears a lot of caches. It should probably clear these cached statements as well.